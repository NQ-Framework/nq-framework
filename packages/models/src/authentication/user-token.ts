export interface UserToken {
  idToken: string;
  refreshToken: string;
  email: string;
  expiresIn: string;
  localId: string;
  registered: boolean;
}
