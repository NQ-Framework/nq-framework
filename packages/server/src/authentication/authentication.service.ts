import { HttpService, Injectable } from '@nestjs/common';
import { UserToken } from '@nqframework/models';
import { AuthConfigService } from '../config/AuthConfigService';
import { LoggerService } from '../core/logger.service';

@Injectable()
export class AuthenticationService {

    constructor(private logger: LoggerService, private http: HttpService, private authConfig: AuthConfigService) {
        logger.setContext("Authentication Service");
    }

    async loginUserWithEmailAndPassword(email: string, password: string): Promise<UserToken> {
        this.logger.debug(`Logging in user ${email} with password`);
        try {
            const result = await this.http.post(`https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=${this.authConfig.webApiKey}`, {
                email,
                password,
                returnSecureToken: true
            }, {
                headers: {
                    'content-type': 'application/json'
                }
            }).toPromise();
            this.logger.debug(`successful authentication for ${email}`)
            return result.data;
        }
        catch (error) {
            this.logger.error(`failed logging in user ${email}`, error);
            throw new Error(error);
        }
    }
}
