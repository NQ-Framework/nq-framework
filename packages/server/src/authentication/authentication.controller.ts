import { BadRequestException, Body, Controller, Post } from '@nestjs/common';
import { AuthenticationService } from './authentication.service';

@Controller('authentication')
export class AuthenticationController {
  constructor(private service: AuthenticationService) {}

  @Post('user')
  async UserLogin(
    @Body('email') email: string,
    @Body('password') password: string,
  ) {
    if (!email || !password) {
      throw new BadRequestException('Missing user credentials in body');
    }
    const token = await this.service.loginUserWithEmailAndPassword(
      email,
      password,
    );
    return {
      email: token.email,
      idToken: token.idToken,
      refreshToken: token.refreshToken,
      expiresIn: token.expiresIn,
      localId: token.localId,
    };
  }

  @Post('token')
  async ServiceAccountLogin(
    @Body('client_id') email: string,
    @Body('client_secret') password: string,
  ) {
    if (!email || !password) {
      throw new BadRequestException('Invalid client credentials in body');
    }
    const token = await this.service.loginUserWithEmailAndPassword(
      email,
      password,
    );
    return {
      idToken: token.idToken,
      refreshToken: token.refreshToken,
      expiresIn: token.expiresIn,
    };
  }
}
