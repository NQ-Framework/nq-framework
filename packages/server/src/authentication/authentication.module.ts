import { HttpModule, Module } from '@nestjs/common';
import { CoreModule } from '../core/core.module';
import { ConfiugrationModule } from '../config/configuration.module';
import { AuthenticationController } from './authentication.controller';
import { AuthenticationService } from './authentication.service';

@Module({
  imports: [HttpModule, ConfiugrationModule, CoreModule],
  controllers: [AuthenticationController],
  providers: [AuthenticationService]
})
export class AuthenticationModule { }
