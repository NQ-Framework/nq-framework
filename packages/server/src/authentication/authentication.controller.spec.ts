import { Test, TestingModule } from '@nestjs/testing';
import { AuthenticationController } from './authentication.controller';
import { AuthenticationService } from './authentication.service';

describe('AuthenticationController', () => {
  let controller: AuthenticationController;
  let authService: AuthenticationService;
  const mockToken = {
    email: "mock@mock",
    idToken: "mocktoken",
    refreshToken: "mockrefresh",
    expiresIn: "mockexpires",
    localId: "mockid"
  };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [AuthenticationController],
      providers: [
        { provide: AuthenticationService, useValue: { loginUserWithEmailAndPassword: jest.fn() } }
      ]
    }).compile();

    controller = module.get<AuthenticationController>(AuthenticationController);
    authService = module.get<AuthenticationService>(AuthenticationService);
  });

  it('should reject login calls if no input parameters', async () => {
    await expect(controller.UserLogin('', '')).rejects.toThrowErrorMatchingSnapshot();
    await expect(controller.ServiceAccountLogin('', '')).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should call service and return result for service account', async () => {
    const loginMock = authService.loginUserWithEmailAndPassword as jest.Mock;
    loginMock.mockImplementation(() => {
      return mockToken
    });
    const result = await controller.ServiceAccountLogin('mock email', 'mock pw');

    expect(loginMock).toBeCalledWith('mock email', 'mock pw');

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { email, localId, ...expected } = mockToken;
    expect(result).toEqual(expected);
  });

  it('should call service and return result for user account', async () => {
    const loginMock = authService.loginUserWithEmailAndPassword as jest.Mock;
    loginMock.mockImplementation(() => {
      return mockToken
    });
    const result = await controller.UserLogin('mock email', 'mock pw');

    expect(loginMock).toBeCalledWith('mock email', 'mock pw');

    expect(result).toEqual(mockToken);
  });
});
