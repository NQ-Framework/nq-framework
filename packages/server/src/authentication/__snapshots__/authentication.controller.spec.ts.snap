// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticationController should reject login calls if no input parameters 1`] = `"Missing user credentials in body"`;

exports[`AuthenticationController should reject login calls if no input parameters 2`] = `"Invalid client credentials in body"`;
